### DISCUSSION TITLE
Most efficient way to overlay and sjoin 11 Mio. Polygons with Polygon Mask

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [geopandas/geopandas#3063] [Q&A] [answered]
Hello, 

I have two polygon GeoDataFrames. The first one (`gdf1`) contains 11 mio. building footprints and their associated heat demand. The second GeoDataFrame (`gdf2`) contains a polygon grid with 3x3 km² polygons with a total of approx. 4,000 polygons. The goal is to sum up the heat demand of all buildings of `gdf1` in each cell/polygon of `gdf2`. The entire operation currently takes more than 200 h on a pretty decent machine. 

My question is what the fastest way is to process the data? Dask-geopandas? Using the entire `gdf2` at once instead of performing the same operation 4,000 times (see first bullet)? Something else?

What I am doing now is the following (simplified):

* Overlay `gdf1` with one of the polygons of `gdf2` --> `gdf_overlay` (I could probably use the entire `gdf2` at once but I am using list comprehension and the `tqdm` package to track the process) 
* Get the centroid of each resulting polygon 
* Perform spatial join between `gdf_overlay` and the respective polygon of `gdf2` --> `gdf_joined`
* group `gdf_joined` by `index_right` and sum up the values for the heat demand --> `gdf_grouped` 
* Concatenate `gdf_grouped` with `gdf2` to get the final result

The code for the heat demand calculation is here: https://github.com/AlexanderJuestel/pyheatdemand/blob/c186aa83ec161d81a03f42514a9636e223f6eb1a/pyheatdemand/processing.py#L190-L309

**Comment by martinfleis** [comment]
So you want to join data from each building to each grid polygon it intersects?

**Reply by AlexanderJuestel** [reply]
That is correct, yes.

**Reply by theroggy** [reply]
EDIT: oops, I posted the reply at the wrong place in the tree of this thread... so I moved it ;-).

**Reply by AlexanderJuestel** [reply]
@theroggy, 
I may give it a try!

**Reply by martinfleis** [reply]
If you want just a join you don't need an overlay or any expensive intersection. In your case, you can use a spatial index without  predicate to do the join. That will be the most efficient. I'll try to come up with some code tomorrow.

**Reply by AlexanderJuestel** [reply]
The reason for the overlay is that a building footprint may be located within two adjacent cells. So by using overlay, we cut all the building polygons with the mask prior to spatially joining them with each mask polygon. If some spatial joining method can take care of that step automatically, that would be amazing :)

**Reply by martinfleis** [reply]
That doesn't matter, spatial join deals with that.

**Reply by theroggy** [reply]
Hmm... I wasn't aware of a functionality in geopandas that does this in one step...

**Reply by AlexanderJuestel** [reply]
> That doesn't matter, spatial join deals with that. 

I mean, that would be amazing! Looking forward to your working example then

**Reply by martinfleis** [reply, marked as answer]
This is, in principle, the core.

```py
grid_ix, buildings_ix = buildings.sindex.query(grid.geometry, predicate=None)
heat_per_grid_cell = buildings["heat_demand"].iloc[buildings_ix].groupby(grid_ix).sum() # or mean or whatever you want
grid["heat_demand"] = heat_per_grid_cell.values
```

Since you are joining convex building geometries to a grid, you can use `predicate=None` which checks only the intersection of bounding boxes in the STRtree spatial index. That is significantly cheaper than checking for `predicate="intersects`. You can then use the index coming from the query to get heat demand values from each building and group them by grid indices indicating which grid ID each hit belongs to. And then you assing the values to a column. 

Sindex.query gives you all hits, so if a building intersects two grid cells, you get both. 

It still uses a single core but it should not take a ton of time. I suppose minutes, max tens of minutes. Eventually, you can split buildings into parts and apply this snippet in parallel but I am not sure it is worth the hassle as you'd need to be more mindful about memory.

**Reply by AlexanderJuestel** [reply]
Dear @martinfleis, thank you very much! I will give it a try on a little sample data set and will then try it on the full dataset. Let's see how that goes.

**Reply by AlexanderJuestel** [reply]
@martinfleis How is that possible? :D I just had to adapt the last line a little as not all grid polygons contain heat values but what the hell (sorry for the wording). For my test dataset (which took me 10 minutes to process), it only takes about 80 ms!!! This is crazy! Wow

**Reply by martinfleis** [reply]
Happy to help :).

**Reply by AlexanderJuestel** [reply]
@martinfleis implementing your solution for my 11 Mio. polygons takes less than one second now! Down from 240 h, crazy

**Reply by martinfleis** [reply]
That's actually even faster than what we would expect. Great it has worked out!

**Comment by theroggy** [comment]
> My question is what the fastest way is to process the data? Dask-geopandas?

Dask-geopandas doesn't support overlay operations yet... so that's not an option for this use case.

> Using the entire gdf2 at once instead of performing the same operation 4,000 times (see first bullet)?

Doing the overlay in one go instead of per row should be significantly faster. Anyway I wouldn't have suspected this operation to take 200 hours...

> Something else?

If you would like to have a very fast solution that uses parallel processing, and if don't mind trying another library, you could try [geofileops](https://github.com/geofileops/geofileops) (disclaimer: I'm the developer of geofilops). It is able to use all you CPU's to calculate the joins and intersections which should speed things up considerably.

First thing to note is that geofileops works on (.gpkg) files, not on dataframes. So you'll need to save the input data as a .gpkg file first (or you don't need to read the data first).

Untested and just to give an idea on how the code might look:
```
import logging
import geofileops as gfo

if __name__ == "__main__":
    # Init logging so progress printed by gfo is shown
    logging.basicConfig(level=logging.INFO)

    # Write input to file if it only exists in memory
    gdf1_path = "/path/you/like/gdf1.gpkg"
    gdf2_path = "/path/you/like/gdf2.gpkg"
    gfo.to_file(gdf1, gdf1_path)
    gfo.to_file(gdf2, gdf2_path)

    # Calculate intersection
    # Input columns are also joined: columns of gdf1 are prefixed with "l1_", those of gdf2 with "l2_"
    output_path = "/path/you/like/output.gpkg"
    gfo.intersection(input1_path=gdf1_path, input2_path=gdf2_path, output_path=output_path)

    # I am more used to use SQL to calculate statistics... but obviously this can be replaced by using pandas
    statistics_sql = """
        SELECT l2_column_to_group_by
              ,SUM((l1_area_orig / ST_Area({geometrycolumn})) * l1_heat_demand) as total_heat_demand
          FROM "{input_layer}" layer
         GROUP BY l2_column_to_group_by
    """
    stats_df = gfo.read_file(output_path, sql_stmt=statistics_sql)
```
